src/JITModule.h

```diff
@@ -0,0 +1,114 @@
+#ifndef HALIDE_JIT_MODULE_H
+#define HALIDE_JIT_MODULE_H
+
+/** \file
+ * Defines the struct representing a JIT compiled halide pipeline
+ */
+
+#include "IntrusivePtr.h"
+#include "runtime/HalideRuntime.h"
+
+#include <map>
+
+namespace llvm {
+class Module;
+class Type;
+}
+
+namespace Halide {
+
+struct Target;
+
+namespace Internal {
+
+class JITModuleContents;
+class CodeGen;
+
+struct JITModule {
+    IntrusivePtr<JITModuleContents> jit_module;
+
+    struct Symbol {
+        void *address;
+        llvm::Type *llvm_type;
+        Symbol() : address(NULL), llvm_type(NULL) {}
+        Symbol(void *address, llvm::Type *llvm_type) : address(address), llvm_type(llvm_type) {}
+    };
+
+    JITModule() {}
+    JITModule(const std::map<std::string, Symbol> &exports);
+
+    const std::map<std::string, Symbol> &exports() const;
+
+    /** A pointer to the raw halide function. It's true type depends
+     * on the Argument vector passed to CodeGen::compile. Image
+     * parameters become (buffer_t *), and scalar parameters become
+     * pointers to the appropriate values. The final argument is a
+     * pointer to the buffer_t defining the output. This will be NULL for
+     * a JITModule which has not yet been compiled or one that is not
+     * a Halide Func compilation at all. */
+    void *main_function() const;
+
+    /** A slightly more type-safe wrapper around the raw halide
```

Comment:
I'd say this is *less* type safe, not more. The advantage of this function is that it is easier for a program to call, because it is difficult to generate a call site for argument types that aren't known at compile time. This function allows you to build an array at runtime to call with.